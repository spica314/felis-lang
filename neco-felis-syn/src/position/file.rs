use crate::{Item, Parse, ParseError, Phase, PhaseParse, token::Token};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct File<P: Phase> {
    pub items: Vec<Item<P>>,
    pub ext: P::FileExt,
}

impl<P: Phase> File<P> {
    pub fn items(&self) -> &[Item<P>] {
        &self.items
    }
}

impl Parse for File<PhaseParse> {
    fn parse(tokens: &[Token], i: &mut usize) -> Result<Option<Self>, ParseError> {
        let mut k = *i;

        let mut items = vec![];
        while let Some(item) = Item::parse(tokens, &mut k)? {
            items.push(item);
        }

        let file = File { items, ext: () };

        *i = k;
        Ok(Some(file))
    }
}

#[cfg(test)]
mod test {
    use crate::{FileIdGenerator, token::Token};

    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse_inductive_eq() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_eq.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_inductive_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/inductive_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_eq_and_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/eq_and_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_exit_42() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/exit_42.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());
    }

    #[test]
    fn test_parse_let_mut() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/let_mut.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let result = File::parse(&tokens, &mut i);

        match result {
            Ok(Some(file)) => {
                println!("Parsed successfully up to token {i} of {}", tokens.len());
                if i < tokens.len() {
                    println!("Remaining tokens: {:?}", &tokens[i..]);
                }
                assert_debug_snapshot!(file);
                assert_eq!(i, tokens.len());
            }
            Ok(None) => {
                println!("Parsing returned None at token {i} of {}", tokens.len());
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parsing returned None");
            }
            Err(e) => {
                println!("Parse error at token {i} of {}: {:?}", tokens.len(), e);
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parse error: {e:?}");
            }
        }
    }

    #[test]
    fn test_parse_array() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/array.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let result = File::parse(&tokens, &mut i);

        match result {
            Ok(Some(file)) => {
                println!("Parsed successfully up to token {i} of {}", tokens.len());
                if i < tokens.len() {
                    println!("Remaining tokens: {:?}", &tokens[i..]);
                }
                assert_debug_snapshot!(file);

                // Complete file parsing including #proc block is now working
                assert_eq!(i, tokens.len()); // Should have parsed all tokens

                // Verify the array was parsed correctly - it should be the 6th item (0-indexed)
                assert_eq!(file.items.len(), 7);
                match &file.items[5] {
                    crate::Item::Array(array) => {
                        assert_eq!(array.name().s(), "Points");
                        assert_eq!(array.fields().len(), 2);
                    }
                    _ => panic!("Expected 5th item to be an array"),
                }
            }
            Ok(None) => {
                println!("Parsing returned None at token {i} of {}", tokens.len());
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parsing returned None");
            }
            Err(e) => {
                println!("Parse error at token {i} of {}: {:?}", tokens.len(), e);
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parse error: {e:?}");
            }
        }
    }

    #[test]
    fn test_parse_struct_1() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/struct_1.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        let mut i = 0;
        let file = File::parse(&tokens, &mut i).unwrap().unwrap();

        assert_debug_snapshot!(file);
        assert_eq!(i, tokens.len());

        // Verify the struct was parsed correctly
        // Look for the struct item in the parsed items
        let mut found_struct = false;
        for item in &file.items {
            if let crate::Item::Struct(struct_item) = item {
                assert_eq!(struct_item.name().s(), "Vec3");
                assert_eq!(struct_item.fields().len(), 3);
                assert_eq!(struct_item.fields()[0].name.s(), "x");
                assert_eq!(struct_item.fields()[1].name.s(), "y");
                assert_eq!(struct_item.fields()[2].name.s(), "z");
                found_struct = true;
                break;
            }
        }
        assert!(found_struct, "Vec3 struct not found in parsed items");
    }

    #[test]
    fn test_parse_array_len() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/array_len.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        println!("Total tokens: {}", tokens.len());
        for (i, token) in tokens.iter().enumerate() {
            println!("{i}: {token:?}");
        }

        let mut i = 0;
        let result = File::parse(&tokens, &mut i);

        // Debug: Print status
        if i >= 44 {
            println!("Stopped at token {i}, trying to parse remaining tokens as ItemProc");
        }

        match result {
            Ok(Some(file)) => {
                println!("Parsed successfully up to token {i} of {}", tokens.len());
                if i < tokens.len() {
                    println!("Remaining tokens: {:?}", &tokens[i..]);
                }
                assert_debug_snapshot!(file);

                // Complete file parsing
                assert_eq!(i, tokens.len()); // Should have parsed all tokens
            }
            Ok(None) => {
                println!("Parsing returned None at token {i} of {}", tokens.len());
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parsing returned None");
            }
            Err(e) => {
                println!("Parse error at token {i} of {}: {:?}", tokens.len(), e);
                println!(
                    "Tokens around position: {:?}",
                    &tokens[i.saturating_sub(3)..tokens.len().min(i + 3)]
                );
                panic!("Parse error: {e:?}");
            }
        }
    }
}
