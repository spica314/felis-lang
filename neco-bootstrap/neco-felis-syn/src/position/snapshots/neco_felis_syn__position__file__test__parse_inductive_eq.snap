---
source: neco-felis-syn/src/position/file.rs
expression: file
---
File {
    items: [
        Inductive(
            ItemInductive {
                keyword_inductive: TokenKeyword {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 1,
                    },
                    s: "inductive",
                },
                name: TokenVariable {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 12,
                    },
                    s: "eq",
                },
                colon: TokenColon {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 15,
                    },
                },
                ty: ArrowDep(
                    TermArrowDep {
                        paren_l: TokenParenL {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 1,
                                column: 17,
                            },
                        },
                        from: TermVariable {
                            variable: TokenVariable {
                                pos: Pos {
                                    file_id: FileId(
                                        0,
                                    ),
                                    line: 1,
                                    column: 18,
                                },
                                s: "A",
                            },
                            ext: (),
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 1,
                                column: 20,
                            },
                        },
                        from_ty: Variable(
                            TermVariable {
                                variable: TokenVariable {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 22,
                                    },
                                    s: "Set",
                                },
                                ext: (),
                            },
                        ),
                        paren_r: TokenParenR {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 1,
                                column: 25,
                            },
                        },
                        arrow: TokenOperator {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 1,
                                column: 27,
                            },
                            s: "->",
                            after_whitespace: true,
                        },
                        to: ArrowDep(
                            TermArrowDep {
                                paren_l: TokenParenL {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 30,
                                    },
                                },
                                from: TermVariable {
                                    variable: TokenVariable {
                                        pos: Pos {
                                            file_id: FileId(
                                                0,
                                            ),
                                            line: 1,
                                            column: 31,
                                        },
                                        s: "x",
                                    },
                                    ext: (),
                                },
                                colon: TokenColon {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 33,
                                    },
                                },
                                from_ty: Variable(
                                    TermVariable {
                                        variable: TokenVariable {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 1,
                                                column: 35,
                                            },
                                            s: "A",
                                        },
                                        ext: (),
                                    },
                                ),
                                paren_r: TokenParenR {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 36,
                                    },
                                },
                                arrow: TokenOperator {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 1,
                                        column: 38,
                                    },
                                    s: "->",
                                    after_whitespace: true,
                                },
                                to: ArrowNodep(
                                    TermArrowNodep {
                                        from: Variable(
                                            TermVariable {
                                                variable: TokenVariable {
                                                    pos: Pos {
                                                        file_id: FileId(
                                                            0,
                                                        ),
                                                        line: 1,
                                                        column: 41,
                                                    },
                                                    s: "A",
                                                },
                                                ext: (),
                                            },
                                        ),
                                        arrow: TokenOperator {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 1,
                                                column: 43,
                                            },
                                            s: "->",
                                            after_whitespace: true,
                                        },
                                        to: Variable(
                                            TermVariable {
                                                variable: TokenVariable {
                                                    pos: Pos {
                                                        file_id: FileId(
                                                            0,
                                                        ),
                                                        line: 1,
                                                        column: 46,
                                                    },
                                                    s: "Prop",
                                                },
                                                ext: (),
                                            },
                                        ),
                                        ext: (),
                                    },
                                ),
                                ext: (),
                            },
                        ),
                        ext: (),
                    },
                ),
                brace_l: TokenBraceL {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 1,
                        column: 51,
                    },
                },
                branches: [
                    ItemInductiveBranch {
                        name: TokenVariable {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 5,
                            },
                            s: "eq_refl",
                        },
                        colon: TokenColon {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 12,
                            },
                        },
                        ty: ArrowDep(
                            TermArrowDep {
                                paren_l: TokenParenL {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 2,
                                        column: 14,
                                    },
                                },
                                from: TermVariable {
                                    variable: TokenVariable {
                                        pos: Pos {
                                            file_id: FileId(
                                                0,
                                            ),
                                            line: 2,
                                            column: 15,
                                        },
                                        s: "A",
                                    },
                                    ext: (),
                                },
                                colon: TokenColon {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 2,
                                        column: 17,
                                    },
                                },
                                from_ty: Variable(
                                    TermVariable {
                                        variable: TokenVariable {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 19,
                                            },
                                            s: "Set",
                                        },
                                        ext: (),
                                    },
                                ),
                                paren_r: TokenParenR {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 2,
                                        column: 22,
                                    },
                                },
                                arrow: TokenOperator {
                                    pos: Pos {
                                        file_id: FileId(
                                            0,
                                        ),
                                        line: 2,
                                        column: 24,
                                    },
                                    s: "->",
                                    after_whitespace: true,
                                },
                                to: ArrowDep(
                                    TermArrowDep {
                                        paren_l: TokenParenL {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 27,
                                            },
                                        },
                                        from: TermVariable {
                                            variable: TokenVariable {
                                                pos: Pos {
                                                    file_id: FileId(
                                                        0,
                                                    ),
                                                    line: 2,
                                                    column: 28,
                                                },
                                                s: "x",
                                            },
                                            ext: (),
                                        },
                                        colon: TokenColon {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 30,
                                            },
                                        },
                                        from_ty: Variable(
                                            TermVariable {
                                                variable: TokenVariable {
                                                    pos: Pos {
                                                        file_id: FileId(
                                                            0,
                                                        ),
                                                        line: 2,
                                                        column: 32,
                                                    },
                                                    s: "A",
                                                },
                                                ext: (),
                                            },
                                        ),
                                        paren_r: TokenParenR {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 33,
                                            },
                                        },
                                        arrow: TokenOperator {
                                            pos: Pos {
                                                file_id: FileId(
                                                    0,
                                                ),
                                                line: 2,
                                                column: 35,
                                            },
                                            s: "->",
                                            after_whitespace: true,
                                        },
                                        to: Apply(
                                            TermApply {
                                                f: Variable(
                                                    TermVariable {
                                                        variable: TokenVariable {
                                                            pos: Pos {
                                                                file_id: FileId(
                                                                    0,
                                                                ),
                                                                line: 2,
                                                                column: 38,
                                                            },
                                                            s: "eq",
                                                        },
                                                        ext: (),
                                                    },
                                                ),
                                                args: [
                                                    Variable(
                                                        TermVariable {
                                                            variable: TokenVariable {
                                                                pos: Pos {
                                                                    file_id: FileId(
                                                                        0,
                                                                    ),
                                                                    line: 2,
                                                                    column: 41,
                                                                },
                                                                s: "A",
                                                            },
                                                            ext: (),
                                                        },
                                                    ),
                                                    Variable(
                                                        TermVariable {
                                                            variable: TokenVariable {
                                                                pos: Pos {
                                                                    file_id: FileId(
                                                                        0,
                                                                    ),
                                                                    line: 2,
                                                                    column: 43,
                                                                },
                                                                s: "x",
                                                            },
                                                            ext: (),
                                                        },
                                                    ),
                                                    Variable(
                                                        TermVariable {
                                                            variable: TokenVariable {
                                                                pos: Pos {
                                                                    file_id: FileId(
                                                                        0,
                                                                    ),
                                                                    line: 2,
                                                                    column: 45,
                                                                },
                                                                s: "x",
                                                            },
                                                            ext: (),
                                                        },
                                                    ),
                                                ],
                                                ext: (),
                                            },
                                        ),
                                        ext: (),
                                    },
                                ),
                                ext: (),
                            },
                        ),
                        comma: TokenComma {
                            pos: Pos {
                                file_id: FileId(
                                    0,
                                ),
                                line: 2,
                                column: 46,
                            },
                        },
                        ext: (),
                    },
                ],
                brace_r: TokenBraceR {
                    pos: Pos {
                        file_id: FileId(
                            0,
                        ),
                        line: 3,
                        column: 1,
                    },
                },
                ext: (),
            },
        ),
    ],
    ext: (),
}
